flink-yarn/src/main/java/org/apache/flink/yarn/FlinkYarnClient.java

```diff
@@ -0,0 +1,653 @@
+		}
+
+		if(taskManagerMemoryMb > maxRes.getMemory() ) {
+			failSessionDuringDeployment();
+			throw new YarnDeploymentException("The cluster does not have the requested resources for the TaskManagers available!\n"
+					+ "Maximum Memory: " + maxRes.getMemory() + " Requested: "+taskManagerMemoryMb + "MB. " + NOTE);
+		}
+
+
+		int totalMemoryRequired = jobManagerMemoryMb + taskManagerMemoryMb * taskManagerCount;
+		ClusterResourceDescription freeClusterMem = getCurrentFreeClusterResources(yarnClient);
+		if(freeClusterMem.totalFreeMemory < totalMemoryRequired) {
+			failSessionDuringDeployment();
+			throw new YarnDeploymentException("This YARN session requires " + totalMemoryRequired + "MB of memory in the cluster. "
+					+ "There are currently only " + freeClusterMem.totalFreeMemory+"MB available.");
+
+		}
+		if( taskManagerMemoryMb > freeClusterMem.containerLimit) {
+			failSessionDuringDeployment();
+			throw new YarnDeploymentException("The requested amount of memory for the TaskManagers ("+taskManagerMemoryMb+"MB) is more than "
+					+ "the largest possible YARN container: "+freeClusterMem.containerLimit);
+		}
+		if( jobManagerMemoryMb > freeClusterMem.containerLimit) {
+			failSessionDuringDeployment();
+			throw new YarnDeploymentException("The requested amount of memory for the JobManager ("+jobManagerMemoryMb+"MB) is more than "
+					+ "the largest possible YARN container: "+freeClusterMem.containerLimit);
+		}
+
+		// ----------------- check if the requested containers fit into the cluster.
+
+		int[] nmFree = Arrays.copyOf(freeClusterMem.nodeManagersFree, freeClusterMem.nodeManagersFree.length);
+		// first, allocate the jobManager somewhere.
+		if(!allocateResource(nmFree, jobManagerMemoryMb)) {
+			failSessionDuringDeployment();
+			throw new YarnDeploymentException("Unable to find a NodeManager that can fit the JobManager/Application master. " +
+					"The JobManager requires " + jobManagerMemoryMb + "MB. NodeManagers available: "+Arrays.toString(freeClusterMem.nodeManagersFree));
+		}
+		// allocate TaskManagers
+		for(int i = 0; i < taskManagerCount; i++) {
+			if(!allocateResource(nmFree, taskManagerMemoryMb)) {
+				failSessionDuringDeployment();
+				throw new YarnDeploymentException("There is not enough memory available in the YARN cluster. " +
+						"The TaskManager(s) require " + taskManagerMemoryMb + "MB each. " +
+						"NodeManagers available: "+Arrays.toString(freeClusterMem.nodeManagersFree) + "\n" +
+						"After allocating the JobManager (" + jobManagerMemoryMb + "MB) and (" + i + "/" + taskManagerCount + ") TaskManagers, " +
+						"the following NodeManagers are available: " + Arrays.toString(nmFree) );
+			}
+		}
+
+		// ------------------ Prepare Application Master Container  ------------------------------
+
+		// respect custom JVM options in the YAML file
+		final String javaOpts = GlobalConfiguration.getString(ConfigConstants.FLINK_JVM_OPTIONS, "");
```

Comment:
Ah ok. But can we get rid of the GlobalConfiguration anyways? Sooner or later the GlobalConfiguration will be removed completely and thus it would be helpful to already start with it.